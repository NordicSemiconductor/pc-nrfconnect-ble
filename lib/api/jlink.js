/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

const familyDefinitions = {
    NRF51: 'nrf51',
    NRF52: 'nrf52',
    UNKNOWN: 'unknown',
};

const deviceTypeDefinitions = {
    NRF51xxx_xxAA_REV1: 'nRF51',
    NRF51xxx_xxAA_REV2: 'nRF51',
    NRF51xxx_xxAC_REV3: 'nRF51',
    NRF51xxx_xxAA_REV3: 'nRF51',
    NRF51xxx_xxAB_REV3: 'nRF51',
    NRF51802_xxAA_REV3: 'nRF51802',
    NRF52832_xxAA_ENGA: 'nRF52832',
    NRF52832_xxAA_REV1: 'nRF52832',
    NRF52832_xxAA_REV2: 'nRF52832',
    NRF52832_xxAA_ENGB: 'nRF52832',
    NRF52840_xxAA_ENGA: 'nRF52840',
    NRF52832_xxAA_FUTURE: 'nRF52832',
    NRF52840_xxAA_FUTURE: 'nRF52840',
    NRF52810_xxAA_REV1: 'nRF52810',
    NRF52810_xxAA_FUTURE: 'nRF52810',
    NRF52832_xxAB_REV1: 'nRF52832',
    NRF52832_xxAB_FUTURE: 'nRF52832',
    NRF51801_xxAB_REV3: 'nRF51801',
};

export const getJlinkDeviceInfo = ({ jlink }) => ({
    family: familyDefinitions[jlink.deviceFamily] || 'unknown',
    type: deviceTypeDefinitions[jlink.deviceVersion] || 'unknown',
    firmwareString: jlink.jlinkObFirmwareVersion,
});
