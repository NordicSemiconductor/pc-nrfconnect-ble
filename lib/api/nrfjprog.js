/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

const familyDefinitions = {
    NRF51: 'nrf51',
    NRF52: 'nrf52',
    UNKNOWN: 'unknown',
};

const deviceTypeDefinitions = {
    NRF51xxx_xxAA_REV1: 'nRF51',
    NRF51xxx_xxAA_REV2: 'nRF51',
    NRF51xxx_xxAC_REV3: 'nRF51',
    NRF51xxx_xxAA_REV3: 'nRF51',
    NRF51xxx_xxAB_REV3: 'nRF51',
    NRF51802_xxAA_REV3: 'nRF51802',
    NRF52832_xxAA_ENGA: 'nRF52832',
    NRF52832_xxAA_REV1: 'nRF52832',
    NRF52832_xxAA_ENGB: 'nRF52832',
    NRF52840_xxAA_ENGA: 'nRF52840',
    NRF52832_xxAA_FUTURE: 'nRF52832',
    NRF52840_xxAA_FUTURE: 'nRF52840',
    NRF52810_xxAA_REV1: 'nRF52810',
    NRF52810_xxAA_FUTURE: 'nRF52810',
    NRF52832_xxAB_REV1: 'nRF52832',
    NRF52832_xxAB_FUTURE: 'nRF52832',
    NRF51801_xxAB_REV3: 'nRF51801',
};

/**
 * Get J-Link device information for the given serial number. The returned
 * promise resolves with an object on the form:
 * {
 *   family: 'nrf52',
 *   type: 'nRF52832',
 *   firmwareString: 'J-Link OB-SAM3U128-V2-NordicSemi...',
 * }
 *
 * @param {Number|String} serialNumber The serial number to get info for.
 * @returns {Promise<Object>} Resolves with object as described above.
 */
// eslint-disable-next-line import/prefer-default-export
export async function getJlinkDeviceInfo({ jlink }) {
    return new Promise(resolve => {
        resolve({
            family: familyDefinitions[jlink.deviceFamily] || 'unknown',
            type: deviceTypeDefinitions[jlink.deviceVersion] || 'unknown',
            firmwareString: jlink.jlinkObFirmwareVersion,
        });
    });
}
