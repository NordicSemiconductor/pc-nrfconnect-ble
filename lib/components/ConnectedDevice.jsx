/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

'use strict';

import React, { PropTypes } from 'react';
import ReactDOM from 'react-dom';

import { Dropdown, MenuItem, Button } from 'react-bootstrap';

import { Connector } from './Connector';

import dfuIcon from '../../resources/dfu_icon.png';

const WINDOW_WIDTH_OFFSET = 375;
const THROTTLE_TIMEOUT = 100;

export default class ConnectedDevice extends React.PureComponent {
    constructor(props) {
        super(props);
        this.boundResizeListener = this._resizeThrottler.bind(this);
    }

    componentDidMount() {
        window.addEventListener('resize', this.boundResizeListener);
        const domNode = ReactDOM.findDOMNode(this);
        if (domNode) {
            this.boundingRect = domNode.getBoundingClientRect();
        }
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.boundResizeListener);
    }

    _resizeThrottler() {
        if (this.resizeTimeout) {
            return;
        }

        this.resizeTimeout = setTimeout(() => {
            this.resizeTimeout = null;
            this._onResize();
        }, THROTTLE_TIMEOUT);
    }

    _onResize() {
        if (!this.boundingRect) {
            return;
        }

        const isCurrentlyBelow = window.innerWidth < (this.boundingRect.right + WINDOW_WIDTH_OFFSET);
        const hasChanged = isCurrentlyBelow !== this.belowWidthThreshold;

        if (!hasChanged) {
            return;
        }

        this.belowWidthThreshold = isCurrentlyBelow;
        this.forceUpdate();
    }

    _onSelect(event, eventKey) {
        const {
            onDisconnect,
            onPair,
            onConnectionParamsUpdate,
            device,
        } = this.props;

        switch (eventKey) {
            case 'Disconnect':
                onDisconnect();
                break;
            case 'Update':
                onConnectionParamsUpdate(device);
                break;
            case 'Pair':
                onPair();
                break;
            default:
                console.log('Unknown eventKey received: ' + eventKey);
        }
    }

    render() {
        const {
            device,
            id,
            sourceId,
            layout,
            isDfuSupported,
        } = this.props;

        const role = device.role === 'central' ? 'Central' : 'Peripheral';

        const style = {
            opacity: device.connected === true ? 1.0 : 0.5,
        };

        const pullRight = this.belowWidthThreshold ? true : false;

        return (
            <div id={id} className='device standalone' style={style}>
                <div className='top-bar'>
                    <div className='flag-line'></div>
                </div>

                <div className='device-body text-small' >
                    <div>
                        <div className='pull-right'>
                            {
                                isDfuSupported ?
                                    <Button id='dfuButton' bsStyle='primary' className='btn-nordic btn-xs' title="Start Secure DFU" onClick={this.props.onClickDfu}>
                                        <img src={dfuIcon} className="icon-dfu-button" />
                                    </Button> : null
                            }
                            <Dropdown pullRight={pullRight} id='connectionDropDown' onClick={() => this._onResize()} onSelect={(eventKey, event) => { this._onSelect(event, eventKey); }}>
                                <Dropdown.Toggle noCaret>
                                    <span className='icon-cog' aria-hidden='true' />
                                </Dropdown.Toggle>
                                <Dropdown.Menu pullRight={pullRight}>
                                    <MenuItem id='updateConnectionMenuItem' eventKey='Update'>Update connection...</MenuItem>
                                    <MenuItem id='pairMenuItem' eventKey='Pair'>Pair...</MenuItem>
                                    <MenuItem id='disconnectMenuItem' eventKey='Disconnect'>Disconnect</MenuItem>
                                </Dropdown.Menu>
                            </Dropdown>
                        </div>
                        <div className='role-flag pull-right'>{role}</div>
                        <strong>{device.name ? device.name : '<Unknown>'}</strong>
                    </div>
                    <div className='address-text'>{device.address}</div>
                </div>
                <Connector sourceId={sourceId} targetId={id} device={device} layout={layout} />
            </div>
        );

        // TODO: later on, we must implement a transition of data from device discovery flags
        // TODO: to connected devices.
        //
        // <div className='flag-line'>
        //     {device.services.map((service, index) => {
        //         return (<div key={index} className='device-flag'>{service}</div>);
        //     })}
        // </div>
    }
}

ConnectedDevice.propTypes = {
    id: PropTypes.string.isRequired,
    device: PropTypes.object.isRequired,
    sourceId: PropTypes.string.isRequired,
    layout: PropTypes.string.isRequired,
    isDfuSupported: PropTypes.bool,
    onClickDfu: PropTypes.func,
    onDisconnect: PropTypes.func.isRequired,
    onPair: PropTypes.func.isRequired,
    onConnectionParamsUpdate: PropTypes.func.isRequired,
};
