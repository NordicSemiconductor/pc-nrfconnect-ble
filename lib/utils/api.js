/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

'use strict';

import { List, Map, OrderedMap, Record } from 'immutable';
import { hackApi } from '../actions/coreActionsHack';

const ImmutableAdapterState = Record({
    instanceId: null,
    port: null,
    serialNumber: null,
    available: false,
    scanning: false,
    advertising: false,
    connecting: false,
    address: null,
    name: null,
    firmwareVerison: null,
});

const ImmutableAdapter = Record({
    state: new ImmutableAdapterState(),
    connectedDevices: Map(),
    /* Adapter sub-reducers */
    deviceDetails: null,
    serverSetup: null,
    security: null,
    isServerSetupApplied: false,
});

const ImmutableDevice = Record({
    instanceId: null,
    isConnecting: false, // Used by UI to determine visualize we are connecting
    connected: false,
    address: null,
    addressType: null,
    advType: null,
    flags: List(),
    adData: OrderedMap(),
    name: null,
    role: null,
    minConnectionInterval: null,
    maxConnectionInterval: null,
    slaveLatency: null,
    connectionSupervisionTimeout: null,
    services: List(),
    rssi: null,
    scanResponse: false,
    bonded: false,
    securityMode: null,
    securityLevel: null,
    isExpanded: false,
    ownPeriphInitiatedPairingPending: false,
});

const ImmutableProperties = Record({
    broadcast: false,
    read: false,
    writeWoResp: false,
    write: false,
    notify: false,
    indicate: false,
    authSignedWr: false,
    reliableWr: false,
    wrAux: false,
});

const ImmutableService = Record({
    instanceId: null,
    deviceInstanceId: null,
    uuid: null,
    name: null,
    handle: null,
    expanded: false,
    discoveringChildren: false,
    children: null,
});

const ImmutableCharacteristic = Record({
    instanceId: null,
    serviceInstanceId: null,
    uuid: null,
    name: null,
    declarationHandle: null,
    valueHandle: null,
    value: List(),
    properties: new ImmutableProperties(),
    readPerm: null,
    writePerm: null,
    fixedLength: false,
    maxLength: null,
    expanded: false,
    discoveringChildren: false,
    children: null,
    errorMessage: null,
});

const ImmutableDescriptor = Record({
    instanceId: null,
    characteristicInstanceId: null,
    uuid: null,
    name: null,
    handle: null,
    value: List(),
    readPerm: null,
    writePerm: null,
    fixedLength: false,
    maxLength: null,
    errorMessage: null,
});

export function getInstanceIds(instanceId) {
    const instanceIds = {
        address: null,
        device: null,
        service: null,
        characteristic: null,
        descriptor: null,
    };

    if (!instanceId) {
        return instanceIds;
    }

    const idArray = instanceId.split('.');

    if (idArray.length > 0) {
        instanceIds.address = idArray[0];
    }

    if (idArray.length > 1) {
        instanceIds.device = idArray.slice(0, 2).join('.');
    }

    if (idArray.length > 2) {
        instanceIds.service = idArray.slice(0, 3).join('.');
    }

    if (idArray.length > 3) {
        instanceIds.characteristic = idArray.slice(0, 4).join('.');
    }

    if (idArray.length > 4) {
        instanceIds.descriptor = idArray.slice(0, 5).join('.');
    }

    return instanceIds;
}

export function getImmutableAdapterState(adapterState) {
    return new ImmutableAdapterState({
        instanceId: adapterState.instanceId,
        port: adapterState.port,
        serialNumber: adapterState.serialNumber,
        available: adapterState.available,
        scanning: adapterState.scanning,
        advertising: adapterState.advertising,
        connecting: adapterState.connecting,
        address: adapterState.address,
        name: adapterState.name,
        firmwareVerison: adapterState.firmwareVerison,
    });
}

export function getImmutableAdapter(adapter) {
    return new ImmutableAdapter({
        port: adapter.state.port,
        state: getImmutableAdapterState(adapter.state),
        connectedDevices: Map(),
        /* Adapter sub-reducers */
        deviceDetails: undefined,
        serverSetup: undefined,
        security: undefined,
    });
}

export function getImmutableDevice(device) {
    return new ImmutableDevice({
        instanceId: device.instanceId,
        connected: device.connected,
        address: device.address,
        addressType: device.addressType,
        name: device.name,
        advType: device.advType,
        adData: Map(device.adData),
        flags: List(device.flags),
        role: device.role,
        minConnectionInterval: device.minConnectionInterval,
        maxConnectionInterval: device.maxConnectionInterval,
        slaveLatency: device.slaveLatency,
        connectionSupervisionTimeout: device.connectionSupervisionTimeout,
        services: List(device.services),
        rssi: device.rssi,
        scanResponse: device.scanResponse,
        ownPeriphInitiatedPairingPending: device.ownPeriphInitiatedPairingPending,
    });
}

export function getImmutableProperties(properties) {
    return new ImmutableProperties({
        broadcast: properties.broadcast,
        read: properties.read,
        writeWoResp: properties.writeWoResp || properties.write_wo_resp,
        write: properties.write,
        notify: properties.notify,
        indicate: properties.indicate,
        authSignedWr: properties.authSignedWr || properties.auth_signed_wr,
        reliableWr: properties.reliableWr || properties.reliable_wr,
        wrAux: properties.wrAux ? properties.wrAux : properties.wr_aux,
    });
}

export function getImmutableService(service) {
    return new ImmutableService({
        instanceId: service.instanceId,
        deviceInstanceId: service.deviceInstanceId,
        uuid: service.uuid,
        name: service.name,
        handle: service.startHandle,
        children: service.children,
    });
}

export function getImmutableCharacteristic(characteristic) {
    const properties = characteristic.properties || {};
    return new ImmutableCharacteristic({
        instanceId: characteristic.instanceId,
        serviceInstanceId: characteristic.serviceInstanceId,
        uuid: characteristic.uuid,
        name: characteristic.name,
        declarationHandle: characteristic.declarationHandle,
        valueHandle: characteristic.valueHandle,
        value: List(characteristic.value),
        properties: getImmutableProperties(properties),
        readPerm: characteristic.readPerm,
        writePerm: characteristic.writePerm,
        fixedLength: characteristic.fixedLength,
        maxLength: characteristic.maxLength,
        children: characteristic.children,
    });
}

function getImmutableDescriptorValue(descriptor) {
    const {
        uuid,
        value,
    } = descriptor;

    const descriptorInstanceIds = getInstanceIds(descriptor.instanceId);

    if (descriptorInstanceIds.device === 'local.server' && uuid === '2902') {
        let cccdValue = new Map();
        const keys = Object.keys(value);
        for (let i = 0; i < keys.length; i += 1) {
            const deviceInstanceId = keys[i];
            cccdValue = cccdValue.set(deviceInstanceId, List(value[deviceInstanceId]));
        }

        return cccdValue;
    }

    return List(value);
}

export function getImmutableDescriptor(descriptor) {
    const value = getImmutableDescriptorValue(descriptor);
    return new ImmutableDescriptor({
        instanceId: descriptor.instanceId,
        characteristicInstanceId: descriptor.characteristicInstanceId,
        uuid: descriptor.uuid,
        name: descriptor.name,
        handle: descriptor.handle,
        value,
        readPerm: descriptor.readPerm,
        writePerm: descriptor.writePerm,
        fixedLength: descriptor.fixedLength,
        maxLength: descriptor.maxLength,
    });
}

export default function asImmutable(mutableObject) {
    // NOTE: No idea why instanceof does not work for some types
    if (mutableObject.constructor.name === 'AdapterState') {
        return getImmutableAdapterState(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Adapter) {
        return getImmutableAdapter(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Device) {
        return getImmutableDevice(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Service) {
        return getImmutableService(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Characteristic) {
        return getImmutableCharacteristic(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Descriptor) {
        return getImmutableDescriptor(mutableObject);
    }
    throw new Error('Explode!');
}
