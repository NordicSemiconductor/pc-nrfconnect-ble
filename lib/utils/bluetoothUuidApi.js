/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { readFileSync, writeFile } from 'fs';
import path from 'path';
import { getUserDataDir, logger } from 'pc-nrfconnect-shared';

const BLUETOOTH_NUMBERS_API_DISPLAY_URL = `https://github.com/NordicSemiconductor/bluetooth-numbers-database/tree/master/v1`;
const BLUETOOTH_NUMBERS_API_URL =
    'https://raw.githubusercontent.com/NordicSemiconductor/bluetooth-numbers-database/master/v1/';
const USER_DATA_DIR = getUserDataDir();
const UUID_FILENAMES = {
    services: 'service_uuids.json',
    characteristics: 'characteristic_uuids.json',
    descriptors: 'descriptor_uuids.json',
};

function writeToFile(fileName, data) {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const fileContent = JSON.stringify(data, null, 4);
    writeFile(filePath, fileContent, err => {
        if (err) {
            logger.debug(`An error ocurred creating the file ${err.message}`);
        }
    });
}

function readFromFile(fileName) {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const uuidDefs = readFileSync(filePath);
    return JSON.parse(uuidDefs);
}

function getFromApi(fileName) {
    const url = BLUETOOTH_NUMBERS_API_URL + fileName;
    return fetch(url)
        .then(response => {
            if (response.status !== 200) {
                throw new Error(
                    `Response code was ${response.status} when trying to access ${url} `
                );
            }
            return response.json();
        })
        .then(data => {
            writeToFile(fileName, data);
            return data;
        })
        .catch(err => {
            throw err;
        });
}

export default () => {
    const uuidKeys = Object.keys(UUID_FILENAMES);
    const responses = uuidKeys.map(key => getFromApi(UUID_FILENAMES[key]));
    return Promise.all(responses)
        .then(([services, characteristics, descriptors]) => {
            logger.info(
                `Updated list of uuids with data from ${BLUETOOTH_NUMBERS_API_DISPLAY_URL}`
            );
            return {
                services,
                characteristics,
                descriptors,
            };
        })
        .catch(err => {
            logger.error(`${err}. Falling back to stored data`);
            try {
                const [services, characteristics, descriptors] = uuidKeys.map(
                    key => readFromFile(UUID_FILENAMES[key])
                );
                return {
                    services,
                    characteristics,
                    descriptors,
                };
            } catch (error) {
                return {};
            }
        });
};
