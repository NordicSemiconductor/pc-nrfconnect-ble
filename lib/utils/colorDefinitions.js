/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

'use strict';

import { logger } from 'pc-nrfconnect-shared';

import '../../resources/css/colordefinitions.scss';

const CSS_CLASS_PREFIX = 'colordefinition';
const FALLBACK_COLOR = { r: 255, g: 0, b: 0 };

function initColors(rules) {
    const colors = {};
    rules.forEach(rule => {
        const regex = new RegExp(`.${CSS_CLASS_PREFIX}-(.*)`);
        const key = regex.exec(rule.selectorText)[1];
        colors[key] = rule.style.color;
    });
    return colors;
}

function findColorDefinitionRules() {
    const rules = [];
    for (let i = 0; i < document.styleSheets.length; i += 1) {
        const { cssRules } = document.styleSheets[i];
        for (let j = 0; j < cssRules.length; j += 1) {
            const rule = cssRules[j];
            if (
                rule.selectorText &&
                rule.selectorText.indexOf(CSS_CLASS_PREFIX) !== -1
            ) {
                rules.push(rule);
            }
        }
    }
    return rules;
}

const colors = initColors(findColorDefinitionRules());

export default {
    getColor(color) {
        const rgbString = colors[color];
        if (!rgbString) {
            logger.debug(`Color ${color} is not defined`);
            return FALLBACK_COLOR;
        }
        const rgbArray = rgbString.replace(/[^\d,]/g, '').split(',');
        return {
            r: rgbArray[0],
            g: rgbArray[1],
            b: rgbArray[2],
        };
    },
};
