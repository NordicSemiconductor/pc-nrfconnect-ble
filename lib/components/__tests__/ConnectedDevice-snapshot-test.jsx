/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import renderer from 'react-test-renderer';

import { getImmutableDevice } from '../../utils/api';
import ConnectedDevice from '../ConnectedDevice';

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn().mockImplementation(() => ({
        app: {
            adapter: {
                bleDriver: {
                    adapter: { _bleDriver: { NRF_SD_BLE_API_VERSION: 5 } },
                },
            },
        },
    })),
}));

const device = getImmutableDevice({
    instanceId: 'device-1',
    connected: true,
    address: 'Device Address',
    name: 'Device Name',
});

describe('ConnectedDevice', () => {
    it('should render correctly without DFU support', () => {
        const tree = renderer
            .create(
                <ConnectedDevice
                    device={device}
                    id="connected-device-id"
                    sourceId="source-id"
                    layout="vertical"
                    isDfuSupported={false}
                    onClickDfu={() => {}}
                    onConnectionParamsUpdate={() => {}}
                    onDisconnect={() => {}}
                    onPair={() => {}}
                    connectedDevicesNumber={1}
                    onPhyUpdate={() => {}}
                    onMtuUpdate={() => {}}
                    onDataLengthUpdate={() => {}}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render correctly with DFU support', () => {
        const tree = renderer
            .create(
                <ConnectedDevice
                    device={device}
                    id="connected-device-id"
                    sourceId="source-id"
                    layout="vertical"
                    isDfuSupported
                    onClickDfu={() => {}}
                    onConnectionParamsUpdate={() => {}}
                    onDisconnect={() => {}}
                    onPair={() => {}}
                    connectedDevicesNumber={1}
                    onPhyUpdate={() => {}}
                    onMtuUpdate={() => {}}
                    onDataLengthUpdate={() => {}}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
