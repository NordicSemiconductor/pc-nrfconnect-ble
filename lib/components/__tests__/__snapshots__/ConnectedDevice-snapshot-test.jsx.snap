exports[`ConnectedDevice should render correctly with DFU support 1`] = `
<div
  className="device standalone"
  id="connected-device-id"
  style={
    Object {
      "opacity": 1,
    }
  }>
  <div
    className="top-bar">
    <div
      className="flag-line" />
  </div>
  <div
    className="device-body text-small">
    <div>
      <div
        className="pull-right">
        <button
          className="btn-nordic btn-xs btn btn-primary"
          disabled={false}
          id="dfuButton"
          onClick={[Function]}
          title="Start Secure DFU"
          type="button">
          <img
            className="icon-dfu-button"
            src="test-file-stub" />
        </button>
        <div
          className="dropdown btn-group"
          onClick={[Function]}>
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle btn btn-default"
            disabled={false}
            id="connectionDropDown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button">
            <span
              aria-hidden="true"
              className="icon-cog" />
          </button>
          <ul
            aria-labelledby="connectionDropDown"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="updateConnectionMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Update connection...
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="pairMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Pair...
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="disconnectMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Disconnect
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="role-flag pull-right">
        Peripheral
      </div>
      <strong>
        Device Name
      </strong>
    </div>
    <div
      className="address-text">
      Device Address
    </div>
  </div>
  <div />
</div>
`;

exports[`ConnectedDevice should render correctly without DFU support 1`] = `
<div
  className="device standalone"
  id="connected-device-id"
  style={
    Object {
      "opacity": 1,
    }
  }>
  <div
    className="top-bar">
    <div
      className="flag-line" />
  </div>
  <div
    className="device-body text-small">
    <div>
      <div
        className="pull-right">
        <div
          className="dropdown btn-group"
          onClick={[Function]}>
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle btn btn-default"
            disabled={false}
            id="connectionDropDown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button">
            <span
              aria-hidden="true"
              className="icon-cog" />
          </button>
          <ul
            aria-labelledby="connectionDropDown"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="updateConnectionMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Update connection...
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="pairMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Pair...
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="disconnectMenuItem"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Disconnect
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="role-flag pull-right">
        Peripheral
      </div>
      <strong>
        Device Name
      </strong>
    </div>
    <div
      className="address-text">
      Device Address
    </div>
  </div>
  <div />
</div>
`;
